include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h

```diff
@@ -24,10 +24,228 @@
 #    include "hetero/algorithm_impl_hetero.h" //TODO: for __brick_copy
 #endif
 
+#if _ONEDPL_CPP20_RANGES_PRESENT
+#    include <ranges>
+#    include <functional>
+#    include "algorithm_ranges_impl.h"
+#endif
+
 namespace oneapi
 {
 namespace dpl
 {
+namespace ext
+{
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+namespace ranges
+{
+
+// [alg.foreach]
+
+struct for_each_fn
+{
+    template<typename _ExecutionPolicy, std::ranges::random_access_range _R, typename _Proj = std::identity,
+             std::indirectly_unary_invocable<std::projected<std::ranges::iterator_t<_R>, _Proj>>  _Fun>
+    requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<_ExecutionPolicy>> && std::ranges::sized_range<_R>
+    constexpr auto
+    operator()(_ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj = {}) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__internal::__select_backend(__exec, std::ranges::begin(__r));
```

Comment:
The tag is obtained differently here and in `oneapi::dpl::experimental::ranges` functions.
The later just pass a range instead of an iterator:

> const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec, __rng);

I am not sure if either of the option is incorrect, but it might be better to align them.

Reply:
I tried to do that...
But it is not implementable right now due to we have got ambiguity w/o concepts, because now we have a general parameter pack like this
__select_backend(policy, _IteratorTypes&&...)

Reply:
I do not understand it... Does the ambiguity step from the fact that the new range algorithms have both device and standard policies? Could you describe it?

Reply:
Example:
```C++
template <class... _IteratorTypes>
__serial_tag<__internal::__is_random_access_iterator<_IteratorTypes...>>
__select_backend(oneapi::dpl::execution::unsequenced_policy, _IteratorTypes&&...)
{
    return {};
}
```

There is no any restrictions to `_IteratorTypes` type. We assume that it is some kind of iterator only.

Reply:
Thanks Mike and Sergey. 
So, the select_backend functions cannot digest the ranges. I suppose, select_backend specialization working only with device policies can be removed then, but it is not very important.

Reply:
This is not true, I think. It should be implementable.

The example:
```cpp
template <class... _IteratorTypes>
__serial_tag<std::false_type> __select_backend(oneapi::dpl::execution::sequenced_policy, _IteratorTypes&&...) { return {}; }

#if _ONEDPL_RANGES_PRESENT
template <class... _IteratorTypes> requires std::conjunction_v<std::ranges::range<_IteratorTypes>>
__serial_tag<std::false_type> __select_backend(oneapi::dpl::execution::sequenced_policy, _IteratorTypes&&...) { return {}; }
#endif // _ONEDPL_RANGES_PRESENT
```

I might lie with the syntax a little bit. But the core idea is it make the second overload more constrained, thus more preferable for the compiler during overload resolution. It works if `_IteratorType` is never `std::ranges::range`, which is exactly our case (I hope)

Overall, I think that the comment from Dmitry is good. We need to decide what we are doing to `__select_backend`.

Reply:
As I told "to resolve ambiguity we have to use `concept and/or requires`. 
Actually, I don't understand motivation to have the second version of `select_backend`. It just increases amount of the code. Or give me a strong  argument which advantages we have got in that case.

Reply:
> This is not true, I think. It should be implementable.

I meant that is not implementable w/o concept and requires.

Reply:
The suggestion was rather stick to one of the approaches for consistency: 

1.  Pass a range directly if you have a range.
1.  Pass an iterator everywhere (specifically, in the old range algorithms), and get rid of `__select_backend` accepting ranges, which works only for the device backend: https://github.com/oneapi-src/oneDPL/blob/e3ebb58c1cc97261909c3bb3b47e4b3747be8dde/include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h#L738

It seems the first is unimplementable, then let's follow the second.

Reply:
The first implementation has
```C++
//TODO required correct implementation of this __ranges::__select_backend()
```
So (2) looks like more preferable and this implementation with `TODO` should be deleted in this PR.

Reply:
It seems there is now a 3rd `__select_backend` approach:
https://github.com/oneapi-src/oneDPL/blob/0eb6eb7b260786c0977f339bf4eec88d5105c93a/include/oneapi/dpl/pstl/execution_impl.h#L110

@MikeDvorskiy can the 2nd approach (see the link below) be replaced with the 3rd everywhere? 

https://github.com/oneapi-src/oneDPL/blob/e3ebb58c1cc97261909c3bb3b47e4b3747be8dde/include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h#L738

Reply:
Actually, after the concepts for the Range API have been added, the situation is becoming simpler - we need only policy type check! Due to just random access ranges are allowed we don't need an access type (random/forward) at all.
const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
So, __ranges::__select_backend(exec)  can be implemented via 4 overloads for 4 host policies, that's all.

Update: I added "`const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);`"

Reply:
What do you think about removing `__select_backend(const execution::device_policy<_KernelName>&, _Ranges&&...)` entities and switching to `__select_backend(<some_policy>)` for `oneapi::dpl::experimental::ranges`?

The discussion has started with the concern that the range algorithms use 2 different approaches to get the tag, while it is possible to stick to only one. The problem is still here.

Reply:
We keep experimental ranges API as is, so we have to keep `__ranges:: __select_backend(const execution::device_policy<_KernelName>&, _Ranges&&...)` as is.